rtemstoolkit/elftoolchain/libelf/elf_memory.3
.\" Copyright (c) 2006,2008 Joseph Koshy.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" This software is provided by Joseph Koshy ``as is'' and
.\" any express or implied warranties, including, but not limited to, the
.\" implied warranties of merchantability and fitness for a particular purpose
.\" are disclaimed.  in no event shall Joseph Koshy be liable
.\" for any direct, indirect, incidental, special, exemplary, or consequential
.\" damages (including, but not limited to, procurement of substitute goods
.\" or services; loss of use, data, or profits; or business interruption)
.\" however caused and on any theory of liability, whether in contract, strict
.\" liability, or tort (including negligence or otherwise) arising in any way
.\" out of the use of this software, even if advised of the possibility of
.\" such damage.
.\"
.\" $Id: elf_memory.3 189 2008-07-20 10:38:08Z jkoshy $
.\"
.Dd June 28, 2006
.Os
.Dt ELF_MEMORY 3
.Sh NAME
.Nm elf_memory
.Nd process an ELF or ar(1) archive mapped into memory
.Sh LIBRARY
.Lb libelf
.Sh SYNOPSIS
.In libelf.h
.Ft "Elf *"
.Fn elf_memory "char *image" "size_t size"
.Sh DESCRIPTION
Function
.Fn elf_memory
is used to process an ELF file or
.Xr ar 1
archive whose image is present in memory.
.Pp
Argument
.Ar image
points to the start of the memory image of the file or archive.
Argument
.Ar size
contains the size in bytes of the memory image.
.Pp
The ELF descriptor is created for reading (i.e., analogous to the
use of
.Xr elf_begin 3
with a command argument value of
.Dv ELF_C_READ Ns ).
.Sh RETURN VALUES
Function
.Fn elf_memory
returns a pointer to a new ELF descriptor if successful, or NULL if an
error occurred.
.Pp
The return value may be queried for the file type using
.Xr elf_kind 3 .
.Sh EXAMPLES
To read parse an elf file, use:
.Bd -literal -offset indent
int fd;
void *p;
struct stat sb;
Elf *e;
\&...
if ((fd = open("./elf-file", O_RDONLY)) < 0 ||
    fstat(fd, &sb) < 0 ||
    (p = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, (off_t) 0)) ==
    MAP_FAILED) {
	... handle system error ...
}

if ((e = elf_memory(p, sb.st_size)) == NULL) {
	... handle elf(3) error ...
}
\&... use ELF descriptor "e" here ...
.Ed
.Sh ERRORS
Function
.Fn elf_memory
can fail with the following errors:
.Bl -tag -width "[ELF_E_RESOURCE]"
.It Bq Er ELF_E_ARGUMENT
A NULL value was used for argument
.Ar image
or the value of argument
.Ar sz
was zero.
.It Bq Er ELF_E_HEADER
The header of the ELF object contained an unsupported value in its
.Va e_ident[EI_CLASS]
field.
.It Bq Er ELF_E_HEADER
The header of the ELF object contained an unsupported value in its
.Va e_ident[EI_DATA]
field.
.It Bq Er ELF_E_RESOURCE
An out of memory condition was detected.
.It Bq Er ELF_E_SEQUENCE
Function
.Fn elf_memory
was called before a working version was set using
.Xr elf_version 3 .
.It Bq Er ELF_E_VERSION
The argument
.Ar image
corresponds to an ELF file with an unsupported version.
.El
.Sh SEE ALSO
.Xr elf 3 ,
.Xr elf_begin 3 ,
.Xr elf_end 3 ,
.Xr elf_errno 3 ,
.Xr elf_kind 3 ,
.Xr gelf 3


ChangeLog
2012-09-06	Chris Johns <chrisj@rtems.org>

	* Discontinued.
	Refer to the git logs.

2011-06-09	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/build.py,
	specbuilder/specbuilder/crossgcc.py,
	specbuilder/specbuilder/defaults.py,
	specbuilder/specbuilder/linux.py,
	specbuilder/specbuilder/spec.py: Add CentOS support for older
	Pythons. Add options to build the tools with specific flags.

2011-02-22	Chris Johns <chrisj@rtems.org>

	* specbuilder/sb-versions, specbuilder/specbuilder/version.py:
	New. Print a suitable version message for automatic documentation
	updating from the spec files in CVS.

	* specbuilder/specbuilder/build.py: Add xz support. Add a function
	to return the name of a package.

	* specbuilder/specbuilder/crossgcc.py: Use the build name in the
	tmp path.

	* specbuilder/specbuilder/darwin.py: Add xz support.

	* specbuilder/specbuilder/defaults.py: Add xz support. Add Windows
	and Linux support.

	* specbuilder/specbuilder/setup.py: Reference the correct shell
	opts. Use the shell setup in the version probe command. Fix the
	version check. Add autotools to the list of spec files to
	install.

	* specbuilder/specbuilder/spec.py: Add changelog and configure
	tests. Manage sub-packages better.

	* specbuilder/specbuilder/version.py,
	specbuilder/specbuilder/windows.py: New.

2010-08-31	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/linux.py: Add Linux support. Can be
	useful on those distributions that do not support RPM
	packages. Users should always be encouraged to use the RPM
	packages provided by the project.

	* specbuilder/specbuilder/default.py: Add Linux.

2010-08-29	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/defaults.py: Add default configure
	macro. Add --dry-run to the parsed arguments. Fix the case when
	run on a new host that does not have default support.

2010-08-13	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/defaults.py: Provide a default shell.

	* specbuilder/specbuilder/rtems.py: Provide a default autoconf
	version.

	* specbuilder/specbuilder/setup.py: Use the default setup shell
	and check the version of autoconf needed.

2010-08-13	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/setup.py: Add import for execute.

2010-08-13	Chris Johns <chrisj@rtems.org>

	* specbuilder/specbuilder/rtems.py: New. Move RTEMS specific
	settings into the RTEMS specific file from the host specific file.

	* specbuilder/specbuilder/darwin.py: Move the RTEMS specific
	values out. Automatically determine the number of CPUs and let
	make use them all.

	* specbuilder/specbuilder/defaults.py: Add --no-smp and --rebuild
	to the options. Post process --no-smp to set the SMP make flags to
	nil. Add support for RTEMS specific overrides.

	* specbuilder/specbuilder/build.py: Add a notice that building has
	started. If a download happens you do not know when it has
	finished.

	* specbuilder/specbuilder/setup.py: Add --rtems to point to the
	RTEMS source and copy the spec and diff files. Also allow the user
	to --rebuild them.

2010-08-09	Chris Johns <chrisj@rtems.org>

	* specbuilder/perl.prov, specbuilder/perl.req: New.
	* specbuilder/defaults.py: The config.sub path is a sub-dir of the
	command.
	* specbuilder/setup.py: Fix make bug. Install RPMLIB files when
	creating the directory.
	* specbuilder/spec.py: Add '%check'.

2010-08-09	Chris Johns <chrisj@rtems.org>

	* specbuilder, specbuilder/sb-build, specbuilder/sb-crossgcc,
	specbuilder/sb-setup, specbuilder/sb-specdump,
	specbuilder/sb-status, specbuilder/specbuilder/.cvsignore,
	specbuilder/specbuilder/build.py,
	specbuilder/specbuilder/config.sub,
	specbuilder/specbuilder/crossgcc.py,
	specbuilder/specbuilder/darwin.py,
	specbuilder/specbuilder/defaults.py,
	specbuilder/specbuilder/error.py,
	specbuilder/specbuilder/execute.py,
	specbuilder/specbuilder/log.py, specbuilder/specbuilder/setup.py,
	specbuilder/specbuilder/spec.py,
	specbuilder/specbuilder/status.py: New.


linkers/elftoolchain/libelf/elf_memory.c
/*-
 * Copyright (c) 2006,2008 Joseph Koshy
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>

#include <ar.h>
#include <libelf.h>
#include <string.h>

#include "_libelf.h"

LIBELF_VCSID("$Id: elf_memory.c 189 2008-07-20 10:38:08Z jkoshy $");

Elf *
elf_memory(char *image, size_t sz)
{
	Elf *e;

	if (LIBELF_PRIVATE(version) == EV_NONE) {
		LIBELF_SET_ERROR(SEQUENCE, 0);
		return (NULL);
	}

	if (image == NULL || sz == 0) {
		LIBELF_SET_ERROR(ARGUMENT, 0);
		return (NULL);
	}

	if ((e = _libelf_allocate_elf()) == NULL)
		return (NULL);

	e->e_cmd = ELF_C_READ;
	e->e_rawfile = image;
	e->e_rawsize = sz;

#undef	LIBELF_IS_ELF
#define	LIBELF_IS_ELF(P) ((P)[EI_MAG0] == ELFMAG0 && 		\
	(P)[EI_MAG1] == ELFMAG1 && (P)[EI_MAG2] == ELFMAG2 &&	\
	(P)[EI_MAG3] == ELFMAG3)

	if (sz > EI_NIDENT && LIBELF_IS_ELF(image)) {
		_libelf_init_elf(e, ELF_K_ELF);
		e->e_class = image[EI_CLASS];
		e->e_byteorder = image[EI_DATA];
		e->e_version = image[EI_VERSION];

		if (e->e_version > EV_CURRENT) {
			e = _libelf_release_elf(e);
			LIBELF_SET_ERROR(VERSION, 0);
			return (NULL);
		}

		if ((e->e_byteorder != ELFDATA2LSB && e->e_byteorder !=
 		    ELFDATA2MSB) || (e->e_class != ELFCLASS32 && e->e_class !=
		    ELFCLASS64)) {
			e = _libelf_release_elf(e);
			LIBELF_SET_ERROR(HEADER, 0);
			return (NULL);
		}

	} else if (sz >= SARMAG &&
	    strncmp(image, ARMAG, (size_t) SARMAG) == 0) {
		_libelf_init_elf(e, ELF_K_AR);
		e = _libelf_ar_open(e);
	} else
		_libelf_init_elf(e, ELF_K_NONE);

	return (e);
}
